helix-vcs/src/diff.rs

```diff
@@ -0,0 +1,186 @@
+use std::ops::Range;
+use std::sync::Arc;
+
+use helix_core::Rope;
+use imara_diff::Algorithm;
+use parking_lot::{Mutex, MutexGuard};
+use tokio::sync::mpsc::{unbounded_channel, UnboundedSender};
+use tokio::sync::{Notify, RwLock};
+use tokio::task::JoinHandle;
+
+use crate::diff::worker::DiffWorker;
+
+mod line_cache;
+mod worker;
+
+type RedrawHandle = Arc<(Notify, RwLock<()>)>;
+
+// The order of enum variants is used by the PartialOrd
+// derive macro, DO NOT REORDER
+#[derive(PartialEq, PartialOrd)]
+enum RenderStrategy {
+    Async,
+    SyncWithTimeout,
+    Sync,
```

Comment:
I think they are all still required, as they serve different orhogonal usecases:
* `SyncWithTimeout`: Is the standard strategy used in `apply_impl` this will perform the diff in parallel and block the rendering until it is completed or the timout is expired. If the timeout expires before the diff completes a new redraw will be requsted
* `Sync`: Same as `SyncWithTimeout` but without the timeout (so the rendering always waits for the diff to complete). This will be required for side by side diffing where it would look extremely jarring if the the virtual text updates asynchronously and used instead of `SyncWithTimeout` in `apply_impl`.
* `Async`: An update of the `diff_base` trigger by something externally (for example when something is staged/a commit is updated). This strategy is required because it will not be triggered during the rendering. Therefore it always needs to request a new redraw event. 

Note that I found a slight shortcoming in the locking code that required some small refactors (explained below) and removed this enum in the process but the three cases remain